### ISSUE TITLE
Using django-plans with recurring payments

### ISSUE THREAD

**Issue by amitjindal** [django-getpaid/django-plans#64]
Hi, thanks for a great module. I installed and played with the demo and this works beautifully with what I am trying to do. However I need this for a subscription model where someone who is in a plan is on monthly or yearly payments. So I need to generate invoices for that on the same frequency.

Is there a built-in way to handle such payments?

Thanks
Amit

**Comment by cypreess** [comment]
@amitjindal Let me maybe describe what was the architectural idea behind django-plans and then you will be able to answer yourself if this fits your needs. 

In django-plans each user has object UserPlan that describes which type of plan did he purchased and what is the expiration date of this plan. System is automatically checking when this expiration date happens and sends some mailing before. User is asked to manualy create new order for next period, this will generate Order Confirmation document, and then once paid user gets updated UserPlan with additional expiration time.

There is one extra thing, django-plans have also support different periods for plans, so you can define anything you want like 30 days  or 365 days plan extend.

At this moment there is no support for automatic subscriptions, however I can easily see how this could be implemented as some feature. Here are my tips:
* I would add a field to UserPlan `automatic_renewal = True/False`
* I would add a field to PlanPricing definiton `default_for_automatic_renewal=True/False`
* Add setting `PLANS_AUTOMATIC_RENEWAL=True/False` that would set the default state of the field in UserPlan when new user is created.

And then I would add an extra logic that when it sees that account is just going to be expired it first checks what kind of account it is (automatic renewal off or on) and then decide if its off - send email reminding customer to pay manualy, if its on it will search for PlanPricing that have `default_for_automatic_renewal=True`.

That should do the trick and still be as flexible as possible.

**Comment by amitjindal** [comment]
@cypreess Awesome. Thank you so much. I will try to figure this out and ask if I need help. I am also making changes to handle multi-tenant support since you use User objects in models and in multitenant the user is not in shared database.

Really appreciate your quick reply and help.

Regards
Amit

**Comment by Alir3z4** [comment]
@amitjindal were you able to get anywhere with this ?
Would love to know if you could get it done

**Comment by amitjindal** [comment]
@Alir3z4 Nope. I didn't end up completing this and its still on back burner. It seemed much more complex to handle multi-tenancy with this.

**Comment by Alir3z4** [comment]
Oh, I see.
Thanks for letting us know about the result.

I leave this issue open for future discussion and possible implementation of this future.

**Comment by anuj9196** [comment]
#100 This pull has added `is_recurring` flag to mark orders as recurring and one can easily use this field to do tasks accordingly.

**Comment by PetrDlouhy** [comment]
Hi,

I started with experimental implementation of recurring payments with the [django-paypal](https://github.com/spookylukey/django-paypal).
I have written the code in the [django-plans-paypal](https://github.com/PetrDlouhy/django-plans-paypal) as connector to the `django-plans`.

Currently it doesn't need any support from the `django-plans` to have basic functionality. It simply creates new `Order` (uses the existing for first payment) and calls `order.complete_order()` when the succesfull PayPal IPN notification arrives.

Of course I would need to give the user information about state of his/hers subscribtion, so I will have to add some fields like those suggested by @cypreess and @anuj9196. I will try to make PR from this, when my implementation progress.

**Comment by PetrDlouhy** [comment]
Please review recurring code in PR #106

**Comment by PetrDlouhy** [comment]
Recurring/subscription renewals are a first-class feature now: `RecurringUserPlan` stores the provider token, the `autorenew_account` task (management command `autorenew_accounts`) charges on the `PLANS_AUTORENEW_SCHEDULE` retry ladder, and the behavior is extensively tested (plans/tests/test_autorenew_*.py). See the settings documentation for `PLANS_AUTORENEW_SCHEDULE` and slot bookkeeping. Closing — this predates all of that.
